### PR TITLE
Adding REST Service support to WSO2 Governance Registry

### PR DESCRIPTION
**Pull request by madawas**
[Redmine Issue](https://redmine.wso2.com/issues/3747)

* Implemented a media type handler to save swagger definition files.
* Implemented a swagger processor that will extract data from the swagger definition and populate the REST Service.

### CODE REVIEW

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+import org.wso2.carbon.registry.core.Resource;
+import org.wso2.carbon.registry.core.ResourceImpl;
+import org.wso2.carbon.registry.core.config.RegistryContext;
+import org.wso2.carbon.registry.core.exceptions.RegistryException;
+import org.wso2.carbon.registry.core.jdbc.handlers.RequestContext;
+import org.wso2.carbon.registry.core.utils.RegistryUtils;
+import org.wso2.carbon.registry.extensions.utils.CommonConstants;
+
+import javax.xml.namespace.QName;
+import java.util.*;
```
Comment (Rajith90) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: Remove star imports.

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java [hunk 2]
```diff
@@ -43,6 +41,10 @@
 
 import javax.xml.namespace.QName;
 import javax.xml.stream.XMLStreamException;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
 import java.util.*;
```
Comment (Rajith90) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java]: Remove star imports

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 1 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @Rajith90 Fixed with #218 (04e1517)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java (hunk 2 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java]: @Rajith90 Fixed with #218 (350fc4d)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+	 * Extracts the data from wadl and creates an REST Service registry artifact.
+	 *
+	 * @param wadlElement   wadl content.
+	 * @param wadlName      wadl name.
+	 * @param version       wadl version.
+	 * @param wadlPath      wadl path.
+	 * @return              REST Service element.
+	 */
+	public static OMElement createRestServiceArtifact(OMElement wadlElement, String wadlName, String version, String wadlPath) {
+		if(wadlElement == null) {
```
Comment (mushthaq33) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: Formatting issue

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+		name.setText(serviceName);
+		context.setText("/"+serviceName);
+		apiVersion.setText(version);
+		wadl.setText(wadlPath);
+		OMNamespace wadlNamespace = wadlElement.getNamespace();
+		String wadlNamespaceURI = wadlNamespace.getNamespaceURI();
+		String wadlNamespacePrefix = wadlNamespace.getPrefix();
+		OMElement resourcesElement =
+				wadlElement.getFirstChildWithName(new QName(wadlNamespaceURI, "resources", wadlNamespacePrefix));
+		if(resourcesElement != null) {
```
Comment (mushthaq33) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: Formatting issue

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+	/**
+	 * Saves the REST Service registry artifact created from the imported swagger definition.
+	 *
+	 * @param requestContext        information about current request.
+	 * @param data                  service artifact metadata.
+	 * @throws RegistryException    If a failure occurs when adding the api to registry.
+	 */
+	public static String addServiceToRegistry(RequestContext requestContext, OMElement data) throws RegistryException {
+
+		if(requestContext == null || data == null) {
```
Comment (mushthaq33) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: Formatting issue

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/WADLProcessor.java [hunk 6]
```diff
@@ -155,9 +155,19 @@ public String addWadlToRegistry(RequestContext requestContext, Resource resource
...
+	        OMElement serviceElement = RESTServiceUtils.createRestServiceArtifact(wadlElement, wadlName, version,
+	                                                                              RegistryUtils.getRelativePath(
+			                                                                              requestContext
+					                                                                              .getRegistryContext(),
+			                                                                              actualPath));
+	        String servicePath = RESTServiceUtils.addServiceToRegistry(requestContext, serviceElement);
+	        registry.addAssociation(servicePath, actualPath, CommonConstants.DEPENDS);
+	        registry.addAssociation(actualPath, servicePath, CommonConstants.USED_BY);
+	        String endpointPath = createEndpointElement(requestContext, wadlElement, version);
+	        if(endpointPath != null) {
```
Comment (mushthaq33) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/WADLProcessor.java]: Formatting issue

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/WADLProcessor.java [hunk 7]
```diff
@@ -233,10 +243,17 @@ public String importWADLToRegistry(RequestContext requestContext, String commonL
...
-            CommonUtil.addService(serviceElement, requestContext);
+	        OMElement serviceElement = RESTServiceUtils.createRestServiceArtifact(wadlElement, wadlName, version,
+	                                                                              RegistryUtils.getRelativePath(
+			                                                                              requestContext
+					                                                                              .getRegistryContext(),
+			                                                                              actualPath));
+            String servicePath = RESTServiceUtils.addServiceToRegistry(requestContext, serviceElement);
+	        addDependency(servicePath, actualPath);
+			String endpointPath = createEndpointElement(requestContext, wadlElement, version);
+	        if(endpointPath != null) {
```
Comment (mushthaq33) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/WADLProcessor.java]: Formatting issue

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java [hunk 8]
```diff
@@ -1020,4 +1022,56 @@ public static String getPeopleType(OMElement element) { //get rid of this method
...
+		return outputStream;
+	}
+
+	/**
+	 * Closes a given input stream.
+	 *
+	 * @param inputStream   the input steam.
+	 */
+	public static void closeInputStream(InputStream inputStream) {
+		if(inputStream != null) {
```
Comment (mushthaq33) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java]: Formatting issue

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java [hunk 9]
```diff
@@ -1020,4 +1022,56 @@ public static String getPeopleType(OMElement element) { //get rid of this method
...
+		}
+	}
+
+	/**
+	 * Closes a given output stream.
+	 *
+	 * @param outputStream   the output steam.
+	 */
+	public static void closeOutputStream(OutputStream outputStream) {
+		if(outputStream != null) {
```
Comment (mushthaq33) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java]: Formatting issue

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+	 * @param swaggerVersion        swagger version.
+	 * @param resourceObjects       swagger resource object list.
+	 * @return                      The API metadata
+	 * @throws RegistryException    If swagger content is invalid.
+	 */
+	public static OMElement createRestServiceArtifact(JsonObject swaggerDocObject, String swaggerVersion,
+	                                                  String endpointURL, List<JsonObject> resourceObjects)
+			throws RegistryException {
+
+		if(swaggerDocObject == null || swaggerVersion == null) {
```
Comment (mushthaq33) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: Formatting issue

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+		OMNamespace wadlNamespace = wadlElement.getNamespace();
+		String wadlNamespaceURI = wadlNamespace.getNamespaceURI();
+		String wadlNamespacePrefix = wadlNamespace.getPrefix();
+		OMElement resourcesElement =
+				wadlElement.getFirstChildWithName(new QName(wadlNamespaceURI, "resources", wadlNamespacePrefix));
+		if(resourcesElement != null) {
+			String endpointUrl =
+					resourcesElement.getAttributeValue(new QName("base"));
+			endpoint.setText(endpointUrl);
+			if(endpointUrl != null) {
```
Comment (mushthaq33) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: Formatting issue

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/WADLProcessor.java (hunk 6 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/WADLProcessor.java]: @mushthaq33 this code segment is already replaced

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/WADLProcessor.java (hunk 7 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/WADLProcessor.java]: @mushthaq33 this code segment is replaced from ed426b50dffaef29768be3f0df8edcd32f5065a1

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java (hunk 8 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java]: @mushthaq33 Fixed with #218 (48fca14)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java (hunk 9 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/utils/CommonUtil.java]: @mushthaq33 Fixed with #218 (48fca14)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 10 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @mushthaq33 Fixed with #218 (48fca14)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 3 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @mushthaq33 Fixed with #218 (48fca146676a04c678be1a45fbe12b038c11a42e)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 4 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @mushthaq33 Fixed with #218 (48fca14)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 11 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @mushthaq33 this code segment is already replaced.

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 5 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @mushthaq33 Fixed with #218 (48fca14)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 12]
```diff
@@ -0,0 +1,478 @@
...
+	 * @param requestContext        current request information.
+	 * @param endpointElement       endpoint metadata element.
+	 * @param endpointPath          endpoint location.
+	 * @return                      The resource path of the endpoint.
+	 * @throws RegistryException    If fails to add the endpoint to the registry.
+	 */
+	public static String addEndpointToRegistry(RequestContext requestContext, OMElement endpointElement, String endpointPath)
+			throws RegistryException {
+
+		if(requestContext == null || endpointElement == null || endpointPath == null) {
```
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @mushthaq33 Fixed with #218 (48fca14)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/SwaggerMediaTypeHandler.java [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+		String sourceURL = null;
+		InputStream inputStream = null;
+		try {
+			sourceURL = requestContext.getSourceURL();
+
+			if (sourceURL == null || sourceURL.isEmpty()) {
+				throw new RegistryException(CommonConstants.EMPTY_URL);
+			}
+
```
Comment (prasa7) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/SwaggerMediaTypeHandler.java]: Extra blank line

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+	 * @param resourceObjects       swagger resource object list.
+	 * @return                      The API metadata
+	 * @throws RegistryException    If swagger content is invalid.
+	 */
+	public static OMElement createRestServiceArtifact(JsonObject swaggerDocObject, String swaggerVersion,
+	                                                  String endpointURL, List<JsonObject> resourceObjects)
+			throws RegistryException {
+
+		if(swaggerDocObject == null || swaggerVersion == null) {
+			throw new IllegalArgumentException("Arguments are invalid. cannot create the REST service artifact. ");
```
Comment (prasa7) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: Arguments context information is missing.

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+	 * Saves the REST Service registry artifact created from the imported swagger definition.
+	 *
+	 * @param requestContext        information about current request.
+	 * @param data                  service artifact metadata.
+	 * @throws RegistryException    If a failure occurs when adding the api to registry.
+	 */
+	public static String addServiceToRegistry(RequestContext requestContext, OMElement data) throws RegistryException {
+
+		if(requestContext == null || data == null) {
+			throw new IllegalArgumentException("Some or all of the arguments may be null. Cannot add the rest service to registry. ");
```
Comment (prasa7) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: Context information is missing.

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+		endpointPath = getChrootedEndpointLocation(requestContext.getRegistryContext()) + endpointPath;
+
+		String resourceId = endpointResource.getUUID();
+		//set resource UUID
+		resourceId = (resourceId == null) ? UUID.randomUUID().toString() : resourceId;
+
+		endpointResource.setUUID(resourceId);
+		//saving the api resource to repository.
+		registry.put(endpointPath, endpointResource);
+		log.info("Endpoint created at " + endpointPath);
```
Comment (prasa7) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: I think this log should be done from the class which calls this Util class. Is there any special reason to add this info log in this Util class ?

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 17]
```diff
@@ -0,0 +1,478 @@
...
+					httpVerbElement.setText(methodObj.get(SwaggerConstants.METHOD).getAsString());
+
+					//Adding urlPattern element to URITemplate element.
+					uriTemplateElement.addChild(urlPatternElement);
+					uriTemplateElement.addChild(httpVerbElement);
+					uriTemplateElement.addChild(authTypeElement);
+					uriTemplates.add(uriTemplateElement);
+				}
+			}
+
```
Comment (prasa7) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: Extra blank line

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 16 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @prasa7 This info log is already converted to a debug log in 2a77e8a7f0d0bfd6157db1e30ac8783b9eccd3f8

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 14 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @prasa7 Added a more elaborated message with argument context information in 818ebe20b0878325e9f7fddd1b47d83e86f7a34b

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 15 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @prasa7 Added a more elaborated message with argument context information in 818ebe2

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 17 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @prasa7 Fixed with #218 (818ebe2)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/SwaggerMediaTypeHandler.java (hunk 13 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/SwaggerMediaTypeHandler.java]: @prasa7 Fixed with #218 (818ebe2)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 18]
```diff
@@ -0,0 +1,478 @@
...
+		OMElement name = factory.createOMElement(NAME, namespace);
+		OMElement context = factory.createOMElement(CONTEXT, namespace);
+		OMElement apiVersion = factory.createOMElement(VERSION, namespace);
+		OMElement endpoint = factory.createOMElement(ENDPOINT_URL, namespace);
+		OMElement transports = factory.createOMElement(TRANSPORTS, namespace);
+		OMElement wadl = factory.createOMElement(WADL, namespace);
+		List<OMElement> uriTemplates = null;
+
+		provider.setText(CarbonContext.getThreadLocalCarbonContext().getUsername());
+		String serviceName = wadlName.contains(".") ? wadlName.substring(0, wadlName.lastIndexOf(".")) : wadlName;
```
Comment (thushara35) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: use single quotes with lastIndexOf(char), it helps performance.

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 19]
```diff
@@ -0,0 +1,478 @@
...
+	}
+
+	/**
+	 * Saves the REST Service registry artifact created from the imported swagger definition.
+	 *
+	 * @param requestContext        information about current request.
+	 * @param data                  service artifact metadata.
+	 * @throws RegistryException    If a failure occurs when adding the api to registry.
+	 */
+	public static String addServiceToRegistry(RequestContext requestContext, OMElement data) throws RegistryException {
```
Comment (thushara35) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: This is a very complex method. Try to reduce the complexity of this method by moving some generic codes to separate method.
I can see some missing java docs as well.

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java [hunk 20]
```diff
@@ -0,0 +1,478 @@
+/*
```
Comment (thushara35) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: there are lot of commented codes in this class. Please explain why you do so?

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/SwaggerProcessor.java [hunk 21]
```diff
@@ -0,0 +1,369 @@
...
+	private Registry registry;
+	private JsonParser parser;
+	private String swaggerResourcesPath;
+	private String documentVersion;
+	private String endpointUrl;
+	OMElement restServiceElement = null;
+	OMElement endpointElement = null;
+	private String endpointLocation;
+
+	public SwaggerProcessor(RequestContext requestContext) {
```
Comment (thushara35) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/SwaggerProcessor.java]: Please document this public constructor.

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/SwaggerConstants.java [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.carbon.registry.extensions.handlers.utils;
+
+public class SwaggerConstants {
```
Comment (thushara35) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/SwaggerConstants.java]: Please document this public class

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 18 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @thushara35 thanks for the information. Fixed in  #218 (16b43f3)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/SwaggerProcessor.java (hunk 21 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/SwaggerProcessor.java]: @thushara35 documented the constructor in  #218 (16b43f3)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/SwaggerConstants.java (hunk 22 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/SwaggerConstants.java]: @thushara35 documented the public class in  #218 (16b43f3)

File: components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java (hunk 19 above)
Comment (madawas) [inline components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/RESTServiceUtils.java]: @thushara35 I have reduced the complexity of this method by moving a generic code to CommonUtil class and also moving a part of a method to another method in same class. #218 (8a34525824c0891700033092c16b9b7222d27da5)

**Pull request merged**